### DISCUSSION TITLE
Investigate building unified Image Atlas at launch

### DISCUSSION THREAD

**Discussion by EM4Volts** [R2Northstar/Northstar#604] [Ideas]
# Hello frens,

after long consideration i bring forth an idea on something that, sooner or later will need to be handled i think.
Something i call "RIA" - "Runtime Image Atlas"
In order to circumvent the Atlas limit we have.
A Unified image atlas built by Northstar at launch, using ressources from all mods that want to add to said atlas.
This small pitch aswell as thought experiment is structured into 3 parts

- Usecases
- How it would work
- Problems to investigate / find solutions for

## Usecases:
With smaller recent talks on stuff like persistence, rewrites and more people asking about how to mod certain ingame images that are packed into image atlases i believe a solution for the image atlas problem should be investigated.

Image Atlases are needed for:
UI (weapon icons on screen, ui elements)
Calling cards
other RUI stuff

It would allow easy access to custom icons for custom weapons/abilities/titans without having people build their own atlas in a long process which will just fill up a slot in our very limited space of atlases.

## The HOW:

Mod authors would have to add each image as a field inside their mod.json, in a fashion somewhat to what is shown below.
 ![enter image description here](https://private-user-images.githubusercontent.com/87427011/291736248-c86869df-10d7-468a-bbbf-9f02bf2d3061.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMjc5NDYsIm5iZiI6MTcwMzAyNzY0NiwicGF0aCI6Ii84NzQyNzAxMS8yOTE3MzYyNDgtYzg2ODY5ZGYtMTBkNy00NjhhLWJiYmYtOWYwMmJmMmQzMDYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDIzMTQwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3NTk4OThkNjYyMmRmMzMzNjFkZjg2ODViNzZmZjlkNWRkODhiMGFiNzhhNTM3YTQ0OTk1ZGIwY2Y1NWNiOGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tpPi8QcE3jH0_9yu37CCL6DOgO2nitj8MQB084KD2gI)

On launch northstar would now do the following:
Check if a image atlas has been built before, if not > build

Building this Atlas would involve the following:
Get all mods that want to add to said atlas, 
build the atlas image itself (the tiled image consisting of all other images)
converting it.
Authors would supply their images as PNG files, Northstar would convert the final imaget to the DDS file. and build the rpak for it.

## Problems:

Investigate efficient tiling for the images (?preexisting tiling code from mrvn radiant?)
Max directX image size that is allowed? How big, how much before we hit it, what if we hit it?
Build straight to Memory, not even saving as RPAK. (idea: Blue) needs RE

**Comment by ASpoonPlaysGames** [comment]
Converting pngs to dds is a real pain, and also slow. Would be better to force the user to use dds honestly.

**Reply by RoyalBlue1** [reply]
do we know how to combine dds?

**Reply by ASpoonPlaysGames** [reply]
Do we know how to convert png to dds properly without texconv?

**Reply by EM4Volts** [reply]
problem with forcing dds is that dds has multiple compression types, so users have to make sure they use the right one, and certain programs export dds differently / plain wrong at times.
while with png this is more simple

**Reply by GeckoEidechse** [reply]
Hmmm, I wonder if we could add a parser to FlightCore or VTOL or something that can double check that and give instructions on how to fix if it's the wrong compression type.

**Comment by EM4Volts** [comment]
![Screenshot 2023-12-19 033850](https://github.com/R2Northstar/Northstar/assets/87427011/c86869df-10d7-468a-bbbf-9f02bf2d3061)

Image for the idea itself (thanks github cdn)
